rust_porting_cookbook.md

```diff
@@ -0,0 +1,240 @@
+# Intro
+
+The goal of this document is to provide guidance for the developer when
+porting code from Emacs C code to Rust.
+
+## Porting Elisp Primitive Functions: Walk-through
+
+Let's look at porting `numberp` to Rust.
+
+First, make sure you have configured and built Remacs on your
+system. You'll probably want to generate TAGS too, so you can jump to
+definitions of C functions.
+
+This is the definition of `numberp`:
+
+``` c
+DEFUN ("numberp", Fnumberp, Snumberp, 1, 1, 0,
+       doc: /* Return t if OBJECT is a number (floating point or integer).  */
+       attributes: const)
+  (Lisp_Object object)
+{
+  if (NUMBERP (object))
+    return Qt;
+  else
+    return Qnil;
+}
+```
+
+The `DEFUN` macro, in addition to defining a function `Fnumberp`, also
+creates a static struct `Snumberp` that describes the function for Emacs'
+Lisp interpreter.
+
+In Rust, we define a `numberp` function that does the actual work then use
+an attribute (implemented as a procedural macro) named `lisp_fn` that
+handles these definitions for us:
+
+``` rust
+// This is the function that gets called when
+// we call numberp in elisp.
+//
+// `lisp_fn` defines a wrapper function that calls numberp with
+// LispObject values. It also declares a struct that we can pass to
+// defsubr so the elisp interpreter knows about this function.
+
+/// Return t if OBJECT is a number (floating point or integer).
+#[lisp_fn]
+fn numberp(object: LispObject) -> LispObject {
+    LispObject::from_bool(object.is_number())
+}
+```
+
+Due to an issue with procedural macros (#263) `lisp_fn` will make all warnings
+and errors appear to be on its line instead of on the real line of Rust code.
+The easy work around is to comment out `lisp_fn` until the compile succeeds
+then enable it to do a final build and begin testing.
+
+The elisp name of the function is derived from the Rust name, with
+underscores replaced by hyphens.  If that is not possible (like for
+the function `+`), you can give an elisp name as an argument to
+`lisp_fn`, like `#[lisp_fn(name = "+")]`.
+
+Optional arguments are also possible: to make the minimum number of
+arguments from elisp different from the number of Rust arguments,
+pass a `min = "n"` argument.
+
+The docstring of the function should be the same as the docstring
+in the C code.  (Don't wonder about it being a comment there, Emacs
+has some magic that extracts it into a separate file.)
+
```

Comment:
Maybe include a mention of what to do if the function is interactive? That can be an easy thing to miss.

Reply:
There is info later in the document. Is it not enough?

Reply:
Ah I missed that. The info later in the document seems fine.

Reply:
I could link to the section from this location if that would help the reader.